---
title: "Otero-Garcia 2022 ssRNAseq - 01 Prep"
author:
  - name: "Emir Turkes"
date: '`r strftime(Sys.time(), format = "%B %d, %Y")`'
link-citations: true
output:
  html_document:
    code_folding: hide
    number_sections: true
    theme: lumen
    highlight: haddock
    toc: true
    toc_depth: 2
    toc_float:
      collapsed: false
      smooth_scroll: false
knit: (function(inputFile, encoding) {
  rmarkdown::render(
    inputFile, encoding = encoding,
    output_file = file.path(
      "..", "results", "otero_garcia_2022_ssRNAseq_01prep.html"
  ))})
---

<style type="text/css">
body {font-size: 16px;}
h1.title {font-size: 35px;}
h1 {font-size: 24px;}
h2 {font-size: 22px;}
.toc-content {padding-left: 0px; padding-right: 0px;}
div.tocify {width: 100%;}
.tocify-subheader .tocify-item {
  font-size: 0.95em; padding-left: 25px; text-indent: 0;
}
.tocify-subheader .tocify-subheader .tocify-item {
  font-size: 0.95em; padding-left: 35px; text-indent: 0;
}
div.main-container {max-width: none; width: 100%;}
</style>

*This file is a part of [otero-garcia-2022-ssRNAseq](https://github.com/eturkes/otero-garcia-2022-ssRNAseq).*

The table of contents in the top left is clickable and can be used to quickly navigate the document.
To toggle the visibility of code, use the `CODE` toggles at the top right of chunks.
The toggle at the start of the document controls the visibility of all chunks.

```{r}
#    This file is part of otero-garcia-2022-ssRNAseq.
#    Copyright (C) 2024  Emir Turkes, UK DRI at UCL
#
#    This program is free software: you can redistribute it and/or modify
#    it under the terms of the GNU General Public License as published by
#    the Free Software Foundation, either version 3 of the License, or
#    (at your option) any later version.
#
#    This program is distributed in the hope that it will be useful,
#    but WITHOUT ANY WARRANTY; without even the implied warranty of
#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#    GNU General Public License for more details.
#
#    You should have received a copy of the GNU General Public License
#    along with this program.  If not, see <http://www.gnu.org/licenses/>.
#
#    Emir Turkes can be contacted at emir.turkes@eturkes.com

# Load required packages, suppressing startup messages.
# -----------------------------------------------------
library(conflicted)
packages <- c(
  "Seurat", "parallelly", "ggplot2", "dplyr", "ggrepel", "viridis"
)
invisible(
  suppressPackageStartupMessages(
    lapply(packages, FUN = library, character.only = TRUE)
  )
)
# -----------------------------------------------------

# Define global settings.
# -----------------------
knitr::opts_chunk$set(fig.width = 12, fig.height = 7, dpi = 300)
# -----------------------

# Define functions.
# -----------------
source("utils.R")
# -----------------

# Useful variables.
# -----------------
step <- 1

data_dir <- file.path("..", "data")
cache_dir <- file.path("..", "cache", paste0("step", "0", step))
if (!dir.exists(cache_dir)) {
  dir.create(cache_dir, recursive = TRUE)
}

protocol <- c("human", "droplet", "single-cell", "umis")
vars_to_regress <- NULL
parallel_override <- 0
# -----------------

# Metadata to plot after dimensionality reduction and clustering.
# Values in list can include "no_legend and/or "no_label" to exclude those.
# -------------------------------------------------------------------------
metadata_to_plot <- vector("list", length = 11)
names(metadata_to_plot) <- c(
  "seurat_clusters", "Cell.Types", "Sample.ID", "SORT", "donor_id", "Braak",
  "Amyloid", "Age", "sex", "RIN", "PMI"
)
metadata_to_plot$Sample.ID <- "no_label"
metadata_to_plot$SORT <- "no_label"
metadata_to_plot$donor_id <- "no_label"
metadata_to_plot$Braak <- "no_label"
metadata_to_plot$Amyloid <- "no_label"
metadata_to_plot$Age <- "no_label"
metadata_to_plot$sex <- "no_label"
metadata_to_plot$RIN <- "no_label"
metadata_to_plot$PMI <- "no_label"
# -------------------------------------------------------------------------
```

# Cleaning

We first read in the data and form a Seurat object.

```{r}
seurat <- readRDS(file.path(data_dir, "excitatory.rds"))
seurat <- seurat[ , -grep("CTRL", x = seurat$donor_id)]
seurat
```

# Cluster Pipeline

We run a pipeline that applies normalisation and scaling, dimensionality reduction, and clustering.

```{r}
sub_name <- "prep"
seurat <- cluster_pipeline(
  seurat, cache_dir = cache_dir,
  sub_name = sub_name, protocol = protocol,
  vars_to_regress = vars_to_regress, parallel_override = parallel_override,
  cc = FALSE, res_divider = 100000
)
seurat

for (i in seq_along(metadata_to_plot)) {
  if (
    "no_legend" %in% metadata_to_plot[[i]] &&
      "no_label" %in% metadata_to_plot[[i]]
  ) {
    print(
      red_dim_plot(
        seurat, x = "umap1", y = "umap2", color = names(metadata_to_plot)[i]
      ) +
        NoLegend() +
        ggtitle(names(metadata_to_plot)[i])
    )
  } else if ("no_legend" %in% metadata_to_plot[[i]]) {
    print(
      red_dim_plot(
        seurat, x = "umap1", y = "umap2",
        color = names(metadata_to_plot)[i], type = "cat"
      ) +
        NoLegend() +
        ggtitle(names(metadata_to_plot)[i])
    )
  } else if ("no_label" %in% metadata_to_plot[[i]]) {
    print(
      red_dim_plot(
        seurat, x = "umap1", y = "umap2", color = names(metadata_to_plot)[i]
      ) +
        ggtitle(names(metadata_to_plot)[i])
    )
  } else {
    print(
      red_dim_plot(
        seurat, x = "umap1", y = "umap2",
        color = names(metadata_to_plot)[i], type = "cat"
      ) +
        ggtitle(names(metadata_to_plot)[i])
    )
  }
}

if (protocol[4] == "umis") {
  print(
    red_dim_plot(
      seurat, x = "umap1", y = "umap2", color = "nFeature_SCT", type = "cont"
    ) + ggtitle("nFeature_SCT")
  )
  print(
    red_dim_plot(
      seurat, x = "umap1", y = "umap2", color = "nCount_SCT", type = "cont"
    ) + ggtitle("nCount_SCT")
  )
} else if (protocol[4] == "reads") {
  print(
    red_dim_plot(
      seurat, x = "umap1", y = "umap2", color = "nFeature_RNA", type = "cont"
    ) + ggtitle("nFeature_RNA")
  )
  print(
    red_dim_plot(
      seurat, x = "umap1", y = "umap2", color = "nCount_RNA", type = "cont"
    ) + ggtitle("nCount_RNA")
  )
}

red_dim_plot(
  seurat, x = "umap1", y = "umap2", color = "percent.mt", type = "cont"
) + ggtitle("percent.mt")
```

# References

This is the concluding section of the document, where we output the `sessionInfo`, and create a bibliography for works cited.

```{r}
sessionInfo()
```
