---
title: "Otero-Garcia 2022 ssRNAseq - 03 Differential Expression"
author:
  - name: "Emir Turkes"
date: '`r strftime(Sys.time(), format = "%B %d, %Y")`'
link-citations: true
output:
  html_document:
    code_folding: hide
    number_sections: true
    theme: lumen
    highlight: haddock
    toc: true
    toc_depth: 3
    toc_float:
      collapsed: false
      smooth_scroll: false
knit: (function(inputFile, encoding) {
  rmarkdown::render(
    inputFile, encoding = encoding,
    output_file = file.path(
      "..", "results", "otero_garcia_2022_ssRNAseq_03diffexpression.html"
    ))})
---

<style type="text/css">
body {font-size: 16px;}
h1.title {font-size: 35px;}
h1 {font-size: 24px;}
h2 {font-size: 22px;}
h3 {font-size: 20px;}
.toc-content {padding-left: 0px; padding-right: 0px;}
div.tocify {width: 100%;}
.tocify-subheader .tocify-item {
  font-size: 0.95em; padding-left: 25px; text-indent: 0;
}
.tocify-subheader .tocify-subheader .tocify-item {
  font-size: 0.95em; padding-left: 35px; text-indent: 0;
}
.tocify-subheader .tocify-subheader .tocify-subheader .tocify-item {
  font-size: 0.95em; padding-left: 45px; text-indent: 0;
}
div.main-container {max-width: none; width: 100%;}
</style>

*This file is a part of [otero-garcia-2022-ssRNAseq](https://github.com/eturkes/otero-garcia-2022-ssRNAseq).*

The table of contents in the top left is clickable and can be used to quickly navigate the document.
To toggle the visibility of code, use the `CODE` toggles at the top right of chunks.
The toggle at the start of the document controls the visibility of all chunks.

```{r}
#    This file is part of otero-garcia-2022-ssRNAseq.
#    Copyright (C) 2024  Emir Turkes, UK DRI at UCL
#
#    This program is free software: you can redistribute it and/or modify
#    it under the terms of the GNU General Public License as published by
#    the Free Software Foundation, either version 3 of the License, or
#    (at your option) any later version.
#
#    This program is distributed in the hope that it will be useful,
#    but WITHOUT ANY WARRANTY; without even the implied warranty of
#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#    GNU General Public License for more details.
#
#    You should have received a copy of the GNU General Public License
#    along with this program.  If not, see <http://www.gnu.org/licenses/>.
#
#    Emir Turkes can be contacted at emir.turkes@eturkes.com

# Load required packages, suppressing startup messages.
# -----------------------------------------------------
library(conflicted)
packages <- c(
  "GSEABase", "SingleCellExperiment", "scales", "ComplexHeatmap", "DT",
  "circlize", "limma"
)
invisible(
  suppressPackageStartupMessages(
    lapply(packages, FUN = library, character.only = TRUE)
  )
)
# -----------------------------------------------------

# Define global settings.
# -----------------------
knitr::opts_chunk$set(fig.width = 10, fig.height = 7, dpi = 96)
# -----------------------

# Define functions.
# -----------------
source("utils.R")
# -----------------

# Useful variables.
# -----------------
step <- 3

cache_dir <- file.path("..", "cache", paste0("step", "0", step))
if (!dir.exists(cache_dir)) {
  dir.create(cache_dir, recursive = TRUE)
}
# -----------------
```

# Prep

```{r}
gene_sets <- readRDS(
  file.path(cache_dir, "..", "step02", "gene_sets.rds")
)
gene_sets_keep <- readRDS(
  file.path(cache_dir, "..", "step02", "gene_sets_keep.rds")
)
gene_anno <- readRDS(file.path(cache_dir, "..", "step02", "gene_anno.rds"))
gse_data <- readRDS(file.path(cache_dir, "..", "step02", "gse_data.rds"))
deg_data <- readRDS(file.path(cache_dir, "..", "step02", "deg_data.rds"))
gse_list <- readRDS(file.path(cache_dir, "..", "step02", "gse_list.rds"))
deg_list <- readRDS(file.path(cache_dir, "..", "step02", "deg_list.rds"))

map <- setNames(gene_anno$symbol, gene_anno$ensembl)

ctypes <- vector("list", length = 1)
ctypes <- list(
  `L2/4_Cortex` = c("Ex01_CUX2-LAMP5 (L2-L3)", "Ex02_CUX2-COL5A2 (L2-L4)")
)

priority <- c(
  "tau-protein", "tau protein", "neurofi", "microtubule", "dynein", "cytoskele",
  "laminin", "inclusion", "aggrep", "death", "necro", "folding", "folded",
  "synap", "axon", "dendrit", "senescen", "apopto", "exocyt", "endocyt",
  "phagocyt", "proteasom", "lysosom", "endoplasmic", "Golgi", "Hsp",
  "heat shock", "ESCRT", "ubiqui", "heparan", "chondroitin", "sulfotransferase",
  "glycosaminoglycan", "transcription", "translation", "amyloid",
  "apolipoprotein", "lipoprotein", "fatty", "cholesterol", "insulin", "pruning",
  "phosphatidylserine", "complement", "T cell", "B cell", "immun", "inflammat",
  "stress", "cortisol", "interleukin", "leukocyte", "myeloid", "NF-kappaB",
  "toll-like", "macrophage", "astrocyte", "oligodendrocyte", "Schwann", "myelin"
)

anno_color <- hue_pal()(8)
top_anno <- HeatmapAnnotation(
  Donor = gse_data[[1]]$donor_id,
  cell_type = anno_block(
    gp = gpar(fill = "lightgray", col = "lightgray"),
    labels = unique(gse_data[[1]]$tau),
    labels_gp = gpar(col = "black", fontsize = 12),
    height = unit(5, "mm")
  ),
  col = list(
    Donor = c(
      "Subject6" = anno_color[1], "Subject8" = anno_color[2],
      "Subject4" = anno_color[3], "Subject5" = anno_color[4],
      "Subject2" = anno_color[5], "Subject3" = anno_color[6],
      "Subject1" = anno_color[7], "Subject7" = anno_color[8]
    )
  ),
  simple_anno_size = unit(3, "mm"),
  show_annotation_name = FALSE
)
```

# NFT vs. CTRL

```{r}
ctype_idx <- 1

gse <- gse_data[[ctype_idx]]
deg <- deg_data[[ctype_idx]]
gse
deg

gse_corrected <- removeBatchEffect(
  logcounts(gse), batch = gse$donor_id, group = gse$tau
)
deg_corrected <- removeBatchEffect(
  logcounts(deg), batch = deg$donor_id, group = deg$tau
)

gse_range <- c(
  min(gse_corrected),
  (min(gse_corrected) + max(gse_corrected)) / 2,
  max(gse_corrected)
)
deg_range <- c(
  min(deg_corrected),
  (min(deg_corrected) + max(deg_corrected)) / 2,
  max(deg_corrected)
)
```

## Whole Data

### DE Gene Sets

```{r}
design <- model.matrix(~ 0 + gse$tau + gse$donor_id)
colnames(design) <- make.names(colnames(design))
cont_mat <- makeContrasts(
  NFT_CTRL =
    gse.tauNFT -
    gse.tauCTRL,
  levels = design
)

fit <- lmFit(logcounts(gse), design)
fit <- contrasts.fit(fit, cont_mat)
fit <- eBayes(fit, trend = TRUE, robust = TRUE)
plotSA(fit)

tests <- decideTests(fit, method = "global")
write.fit(
  fit, tests, file.path(cache_dir, "results.tsv"),
  adjust = "BH", F.adjust = "BH", method = "global"
)
results <- read.delim(file.path(cache_dir, "results.tsv"))

rownames(results) <- results$X
results <- results[ , -1]
results <- results[results$P.value.adj < 0.05, ]
results <- results[order(results$P.value.adj), ]
results$gene_set <- rownames(results)

results_up <- results[results$Coef > 0, ]
results_down <- results[results$Coef < 0, ]

datatable_download(results)
```

```{r, fig.width = 12, fig.height = 8}
mat <- gse_corrected[
  rownames(gse) %in% rownames(results_up)[1:30], , drop = FALSE
]
mat <- mat[
  match(rownames(results_up)[1:30], rownames(mat)), , drop = FALSE
]
Heatmap(
  mat,
  col = colorRamp2(gse_range, colors = c("blue", "white", "red")),
  column_title = names(ctypes)[ctype_idx],
  cluster_rows = FALSE,
  cluster_columns = FALSE,
  column_split = rep(seq(2), each = 8),
  row_title_gp = gpar(fontsize = 7),
  row_title_rot = 0,
  top_annotation = top_anno,
  show_column_names = FALSE,
  row_names_max_width = max_text_width(rownames(mat)),
  heatmap_legend_param = list(
    title = "logGSE", title_position = "lefttop-rot"
  )
)
mat <- t(
  apply(
    mat, MARGIN = 1,
    FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
  )
)
Heatmap(
  mat,
  col = colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red")),
  column_title = names(ctypes)[ctype_idx],
  cluster_rows = FALSE,
  cluster_columns = FALSE,
  column_split = rep(seq(2), each = 8),
  row_title_gp = gpar(fontsize = 7),
  row_title_rot = 0,
  top_annotation = top_anno,
  show_column_names = FALSE,
  row_names_max_width = max_text_width(rownames(mat)),
  heatmap_legend_param = list(
    title = "Scaled logGSE", title_position = "lefttop-rot"
  )
)

mat <- gse_corrected[
  rownames(gse) %in% rownames(results_down)[1:30], , drop = FALSE
]
mat <- mat[
  match(rownames(results_down)[1:30], rownames(mat)), , drop = FALSE
]
Heatmap(
  mat,
  col = colorRamp2(gse_range, colors = c("blue", "white", "red")),
  column_title = names(ctypes)[ctype_idx],
  cluster_rows = FALSE,
  cluster_columns = FALSE,
  column_split = rep(seq(2), each = 8),
  row_title_gp = gpar(fontsize = 7),
  row_title_rot = 0,
  top_annotation = top_anno,
  show_column_names = FALSE,
  row_names_max_width = max_text_width(rownames(mat)),
  heatmap_legend_param = list(
    title = "logGSE", title_position = "lefttop-rot"
  )
)
mat <- t(
  apply(
    mat, MARGIN = 1,
    FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
  )
)
Heatmap(
  mat,
  col = colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red")),
  column_title = names(ctypes)[ctype_idx],
  cluster_rows = FALSE,
  cluster_columns = FALSE,
  column_split = rep(seq(2), each = 8),
  row_title_gp = gpar(fontsize = 7),
  row_title_rot = 0,
  top_annotation = top_anno,
  show_column_names = FALSE,
  row_names_max_width = max_text_width(rownames(mat)),
  heatmap_legend_param = list(
    title = "Scaled logGSE", title_position = "lefttop-rot"
  )
)
```

### DE Genes

```{r}
design <- model.matrix(~ 0 + deg$tau + deg$donor_id)
colnames(design) <- make.names(colnames(design))
cont_mat <- makeContrasts(
  NFT_CTRL =
    deg.tauNFT -
    deg.tauCTRL,
  levels = design
)

fit <- lmFit(logcounts(deg), design)
fit <- contrasts.fit(fit, cont_mat)
fit <- eBayes(fit, trend = TRUE, robust = TRUE)
plotSA(fit)

tests <- decideTests(fit, method = "global")
write.fit(
  fit, tests, file.path(cache_dir, "results.tsv"),
  adjust = "BH", F.adjust = "BH", method = "global"
)
results <- read.delim(file.path(cache_dir, "results.tsv"))

rownames(results) <- results$X
results <- results[ , -1]
results <- results[results$P.value.adj < 0.05, ]
results <- results[order(results$P.value.adj), ]
gene_anno_sub <- gene_anno[gene_anno$ensembl %in% rownames(results), ]
gene_anno_sub <- gene_anno_sub[
  match(rownames(results), gene_anno_sub$ensembl),
]
results$Gene <- gene_anno_sub$symbol
results <- results[ , c(ncol(results), 2:(ncol(results) -1))]

results_up <- results[results$Coef > 0, ]
results_down <- results[results$Coef < 0, ]

datatable_download(results)
```

```{r, fig.width = 8, fig.height = 8}
mat <- deg_corrected[
  rownames(deg) %in% rownames(results_up)[1:30], , drop = FALSE
]
mat <- mat[
  match(rownames(results_up)[1:30], rownames(mat)), , drop = FALSE
]
rownames(mat) <- results_up$Gene[1:30]
Heatmap(
  mat,
  col = colorRamp2(deg_range, colors = c("blue", "white", "red")),
  column_title = names(ctypes)[ctype_idx],
  cluster_rows = FALSE,
  cluster_columns = FALSE,
  column_split = rep(seq(2), each = 8),
  row_title_gp = gpar(fontsize = 7),
  row_title_rot = 0,
  top_annotation = top_anno,
  show_column_names = FALSE,
  row_names_max_width = max_text_width(rownames(mat)),
  heatmap_legend_param = list(
    title = "logCPM", title_position = "lefttop-rot"
  )
)
mat <- t(
  apply(
    mat, MARGIN = 1,
    FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
  )
)
Heatmap(
  mat,
  col = colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red")),
  column_title = names(ctypes)[ctype_idx],
  cluster_rows = FALSE,
  cluster_columns = FALSE,
  column_split = rep(seq(2), each = 8),
  row_title_gp = gpar(fontsize = 7),
  row_title_rot = 0,
  top_annotation = top_anno,
  show_column_names = FALSE,
  row_names_max_width = max_text_width(rownames(mat)),
  heatmap_legend_param = list(
    title = "Scaled logCPM", title_position = "lefttop-rot"
  )
)

mat <- deg_corrected[
  rownames(deg) %in% rownames(results_down)[1:30], , drop = FALSE
]
mat <- mat[
  match(rownames(results_down)[1:30], rownames(mat)), , drop = FALSE
]
rownames(mat) <- results_down$Gene[1:30]
Heatmap(
  mat,
  col = colorRamp2(deg_range, colors = c("blue", "white", "red")),
  column_title = names(ctypes)[ctype_idx],
  cluster_rows = FALSE,
  cluster_columns = FALSE,
  column_split = rep(seq(2), each = 8),
  row_title_gp = gpar(fontsize = 7),
  row_title_rot = 0,
  top_annotation = top_anno,
  show_column_names = FALSE,
  row_names_max_width = max_text_width(rownames(mat)),
  heatmap_legend_param = list(
    title = "logCPM", title_position = "lefttop-rot"
  )
)
mat <- t(
  apply(
    mat, MARGIN = 1,
    FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
  )
)
Heatmap(
  mat,
  col = colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red")),
  column_title = names(ctypes)[ctype_idx],
  cluster_rows = FALSE,
  cluster_columns = FALSE,
  column_split = rep(seq(2), each = 8),
  row_title_gp = gpar(fontsize = 7),
  row_title_rot = 0,
  top_annotation = top_anno,
  show_column_names = FALSE,
  row_names_max_width = max_text_width(rownames(mat)),
  heatmap_legend_param = list(
    title = "Scaled logCPM", title_position = "lefttop-rot"
  )
)
```

## Pathways of Interest

### DE Gene Sets

```{r}
design <- model.matrix(~ 0 + gse$tau + gse$donor_id)
colnames(design) <- make.names(colnames(design))
cont_mat <- makeContrasts(
  NFT_CTRL =
    gse.tauNFT -
    gse.tauCTRL,
  levels = design
)

fit <- lmFit(logcounts(gse), design)
fit <- contrasts.fit(fit, cont_mat)
fit <- eBayes(fit, trend = TRUE, robust = TRUE)
plotSA(fit)

tests <- decideTests(fit, method = "global")
write.fit(
  fit, tests, file.path(cache_dir, "results.tsv"),
  adjust = "BH", F.adjust = "BH", method = "global"
)
results <- read.delim(file.path(cache_dir, "results.tsv"))

rownames(results) <- results$X
results <- results[ , -1]
results <- results[results$P.value.adj < 0.05, ]
results <- results[order(results$P.value.adj), ]
results$gene_set <- rownames(results)

results <- results[
  unique(unlist(sapply(priority, FUN = grep, x = rownames(results)))),
]
top_gse <- rownames(results)

results_up <- results[results$Coef > 0, ]
results_down <- results[results$Coef < 0, ]

datatable_download(results)
```

```{r, fig.width = 12, fig.height = 8}
mat <- gse_corrected[
  rownames(gse) %in% rownames(results_up)[1:30], , drop = FALSE
]
mat <- mat[
  match(rownames(results_up)[1:30], rownames(mat)), , drop = FALSE
]
Heatmap(
  mat,
  col = colorRamp2(gse_range, colors = c("blue", "white", "red")),
  column_title = names(ctypes)[ctype_idx],
  cluster_rows = FALSE,
  cluster_columns = FALSE,
  column_split = rep(seq(2), each = 8),
  row_title_gp = gpar(fontsize = 7),
  row_title_rot = 0,
  top_annotation = top_anno,
  show_column_names = FALSE,
  row_names_max_width = max_text_width(rownames(mat)),
  heatmap_legend_param = list(
    title = "logGSE", title_position = "lefttop-rot"
  )
)
mat <- t(
  apply(
    mat, MARGIN = 1,
    FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
  )
)
Heatmap(
  mat,
  col = colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red")),
  column_title = names(ctypes)[ctype_idx],
  cluster_rows = FALSE,
  cluster_columns = FALSE,
  column_split = rep(seq(2), each = 8),
  row_title_gp = gpar(fontsize = 7),
  row_title_rot = 0,
  top_annotation = top_anno,
  show_column_names = FALSE,
  row_names_max_width = max_text_width(rownames(mat)),
  heatmap_legend_param = list(
    title = "Scaled logGSE", title_position = "lefttop-rot"
  )
)

mat <- gse_corrected[
  rownames(gse) %in% rownames(results_down)[1:30], , drop = FALSE
]
mat <- mat[
  match(rownames(results_down)[1:30], rownames(mat)), , drop = FALSE
]
Heatmap(
  mat,
  col = colorRamp2(gse_range, colors = c("blue", "white", "red")),
  column_title = names(ctypes)[ctype_idx],
  cluster_rows = FALSE,
  cluster_columns = FALSE,
  column_split = rep(seq(2), each = 8),
  row_title_gp = gpar(fontsize = 7),
  row_title_rot = 0,
  top_annotation = top_anno,
  show_column_names = FALSE,
  row_names_max_width = max_text_width(rownames(mat)),
  heatmap_legend_param = list(
    title = "logGSE", title_position = "lefttop-rot"
  )
)
mat <- t(
  apply(
    mat, MARGIN = 1,
    FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
  )
)
Heatmap(
  mat,
  col = colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red")),
  column_title = names(ctypes)[ctype_idx],
  cluster_rows = FALSE,
  cluster_columns = FALSE,
  column_split = rep(seq(2), each = 8),
  row_title_gp = gpar(fontsize = 7),
  row_title_rot = 0,
  top_annotation = top_anno,
  show_column_names = FALSE,
  row_names_max_width = max_text_width(rownames(mat)),
  heatmap_legend_param = list(
    title = "Scaled logGSE", title_position = "lefttop-rot"
  )
)
```

### DE Genes

```{r}
design <- model.matrix(~ 0 + deg$tau + deg$donor_id)
colnames(design) <- make.names(colnames(design))
cont_mat <- makeContrasts(
  NFT_CTRL =
    deg.tauNFT -
    deg.tauCTRL,
  levels = design
)

fit <- lmFit(logcounts(deg), design)
fit <- contrasts.fit(fit, cont_mat)
fit <- eBayes(fit, trend = TRUE, robust = TRUE)
plotSA(fit)

tests <- decideTests(fit, method = "global")
write.fit(
  fit, tests, file.path(cache_dir, "results.tsv"),
  adjust = "BH", F.adjust = "BH", method = "global"
)
results <- read.delim(file.path(cache_dir, "results.tsv"))

rownames(results) <- results$X
results <- results[ , -1]
results <- results[results$P.value.adj < 0.05, ]
results <- results[order(results$P.value.adj), ]
gene_anno_sub <- gene_anno[gene_anno$ensembl %in% rownames(results), ]
gene_anno_sub <- gene_anno_sub[
  match(rownames(results), gene_anno_sub$ensembl),
]
results$Gene <- gene_anno_sub$symbol
results <- results[ , c(ncol(results), 2:(ncol(results) -1))]

results <- results[
  rownames(results) %in%
    unique(unlist(geneIds(gene_sets[names(gene_sets) %in% top_gse]))),
]

results_up <- results[results$Coef > 0, ]
results_down <- results[results$Coef < 0, ]

datatable_download(results)
```

```{r, fig.width = 8, fig.height = 8}
mat <- deg_corrected[
  rownames(deg) %in% rownames(results_up)[1:30], , drop = FALSE
]
mat <- mat[
  match(rownames(results_up)[1:30], rownames(mat)), , drop = FALSE
]
rownames(mat) <- results_up$Gene[1:30]
Heatmap(
  mat,
  col = colorRamp2(deg_range, colors = c("blue", "white", "red")),
  column_title = names(ctypes)[ctype_idx],
  cluster_rows = FALSE,
  cluster_columns = FALSE,
  column_split = rep(seq(2), each = 8),
  row_title_gp = gpar(fontsize = 7),
  row_title_rot = 0,
  top_annotation = top_anno,
  show_column_names = FALSE,
  row_names_max_width = max_text_width(rownames(mat)),
  heatmap_legend_param = list(
    title = "logCPM", title_position = "lefttop-rot"
  )
)
mat <- t(
  apply(
    mat, MARGIN = 1,
    FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
  )
)
Heatmap(
  mat,
  col = colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red")),
  column_title = names(ctypes)[ctype_idx],
  cluster_rows = FALSE,
  cluster_columns = FALSE,
  column_split = rep(seq(2), each = 8),
  row_title_gp = gpar(fontsize = 7),
  row_title_rot = 0,
  top_annotation = top_anno,
  show_column_names = FALSE,
  row_names_max_width = max_text_width(rownames(mat)),
  heatmap_legend_param = list(
    title = "Scaled logCPM", title_position = "lefttop-rot"
  )
)

mat <- deg_corrected[
  rownames(deg) %in% rownames(results_down)[1:30], , drop = FALSE
]
mat <- mat[
  match(rownames(results_down)[1:30], rownames(mat)), , drop = FALSE
]
rownames(mat) <- results_down$Gene[1:30]
Heatmap(
  mat,
  col = colorRamp2(deg_range, colors = c("blue", "white", "red")),
  column_title = names(ctypes)[ctype_idx],
  cluster_rows = FALSE,
  cluster_columns = FALSE,
  column_split = rep(seq(2), each = 8),
  row_title_gp = gpar(fontsize = 7),
  row_title_rot = 0,
  top_annotation = top_anno,
  show_column_names = FALSE,
  row_names_max_width = max_text_width(rownames(mat)),
  heatmap_legend_param = list(
    title = "logCPM", title_position = "lefttop-rot"
  )
)
mat <- t(
  apply(
    mat, MARGIN = 1,
    FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
  )
)
Heatmap(
  mat,
  col = colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red")),
  column_title = names(ctypes)[ctype_idx],
  cluster_rows = FALSE,
  cluster_columns = FALSE,
  column_split = rep(seq(2), each = 8),
  row_title_gp = gpar(fontsize = 7),
  row_title_rot = 0,
  top_annotation = top_anno,
  show_column_names = FALSE,
  row_names_max_width = max_text_width(rownames(mat)),
  heatmap_legend_param = list(
    title = "Scaled logCPM", title_position = "lefttop-rot"
  )
)
```

# References

This is the concluding section of the document, where we output the `sessionInfo`, and create a bibliography for works cited.

```{r}
sessionInfo()
```
